sdks/python/apache_beam/runners/interactive/display/pcoll_visualization_test.py

```diff
@@ -0,0 +1,152 @@
+    yield [1, 2, 3, 4, 5, 6, 7, 8]
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization._to_element_list', _mock_to_element_list)
+  def test_dynamic_plotting_return_handle(self):
+    h = pv.visualize(self._pcoll, dynamic_plotting_interval=1)
+    self.assertIsInstance(h, timeloop.Timeloop)
+    h.stop()
+
+  @unittest.skipIf(sys.version_info < (3, 5, 3),
+                   'PCollVisualization is not supported on Python 2.')
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization._to_element_list', _mock_to_element_list)
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization.display_facets')
+  def test_dynamic_plotting_update_same_display(self,
+                                                mocked_display_facets):
+    fake_pipeline_result = runner.PipelineResult(runner.PipelineState.RUNNING)
+    ie.current_env().set_pipeline_result(self._p, fake_pipeline_result)
+    # Starts async dynamic plotting that never ends in this test.
+    h = pv.visualize(self._pcoll, dynamic_plotting_interval=0.001)
+    # Blocking so the above async task can execute at least 3 iterations.
+    timeout = time.time() + 1
+    while len(mocked_display_facets.call_args_list) < 3:
+      if time.time() > timeout:
+        break
+    # The first iteration doesn't provide updating_pv to display_facets.
+    _, first_kwargs = mocked_display_facets.call_args_list[0]
+    self.assertEqual(first_kwargs, {})
+    # The following iterations use the same updating_pv to display_facets and so
+    # on.
+    _, second_kwargs = mocked_display_facets.call_args_list[1]
+    updating_pv = second_kwargs['updating_pv']
+    for call in mocked_display_facets.call_args_list[2:]:
+      _, kwargs = call
+      self.assertIs(kwargs['updating_pv'], updating_pv)
+    h.stop()
+
+  @unittest.skipIf(sys.version_info < (3, 6),
+                   'PCollVisualization is not supported on Python 2.')
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization._to_element_list', _mock_to_element_list)
+  @patch('apache_beam.runners.interactive.display.pcoll_visualization'
+         '.PCollVisualization.display_facets')
+  @patch('timeloop.Timeloop.stop')
+  def test_auto_stop_dynamic_plotting_when_job_is_terminated(
+      self,
+      mocked_timeloop,
+      mocked_display_facets):
+    fake_pipeline_result = runner.PipelineResult(runner.PipelineState.RUNNING)
+    ie.current_env().set_pipeline_result(self._p, fake_pipeline_result)
+    # Starts non-stopping async dynamic plotting until the job is terminated.
+    pv.visualize(self._pcoll, dynamic_plotting_interval=0.001)
+    # Blocking so the above async task can execute at least 3 iterations.
+    timeout = time.time() + 1
```

Comment:
This might be replaced with time.sleep(1) to simplify it.

From the while loop and brake, we cannot guarantee the 3 time execution anyway.

Reply:
Sam was against using time.sleep() in test code because they block the test for exactly that amount of time. I don't know if using a while loop really helps with that. Worst case, it also blocks for ~1s and uses a little bit more CPU. Best case, it unblocks faster than time.sleep(1) and doesn't have potential side effects than time.sleep(1) based on the sleep implementation.

It's a tiny bit of trade-off, which one do you recommend? I'm fine with time.sleep(1) though.

Reply:
loop with time.time() > timeout is practically a time.sleep. I agree with Sam that it would be better to not having timing dependencies in tests but this test already has it. Changing from sleep to this version does not address the underlying problem.

I would prefer to simplify:

```
    # Blocking so the above async task can execute at least another 3
    # iterations.
    timeout = time.time() + 1
    while len(mocked_display_facets.call_args_list) < 6:
      if time.time() > timeout:
        break
```
to 
`time.sleep(1)`

Reply:
Simplified it.